python/tank/platform/engine.py

```diff
@@ -1850,51 +1851,99 @@ def _restart_engine(new_context):
     try:
         # Track some of the current state before restarting the engine.
         old_context = engine.context
+        new_context = new_context or engine.context
 
         # Restart the engine. If we were given a new context to use,
         # use it, otherwise restart using the same context as before.
-        new_context = new_context or engine.context
-        current_engine_name = engine.instance_name
-        engine.destroy()
+        with _CoreContextChangeHookGuard(engine.sgtk, old_context, new_context):
+            current_engine_name = engine.instance_name
+            engine.destroy()
 
-        _start_engine(current_engine_name, new_context.tank, old_context, new_context)
+            _start_engine(current_engine_name, new_context.tank, old_context, new_context)
     except TankError, e:
         engine.log_error("Could not restart the engine: %s" % e)
     except Exception:
         engine.log_exception("Could not restart the engine!")
 
 
-def _execute_pre_context_change_hook(tk, current_context, next_context):
+class _CoreContextChangeHookGuard(object):
     """
-    Executes the pre context change hook.
-
-    :param tk: Toolkit instance.
-    :param current_context: Context before the context change.
-    :param next_context: Context after the context change.
+    Used with the ``with`` statement, this guard will notify the context_change
+    core hook with the pre_context_change event on entering and
+    post_context_change even on exit if and only if the scope exits without
+    an exception being raised.
     """
-    tk.execute_core_hook_method(
-        constants.CONTEXT_CHANGE_HOOK,
-        "pre_context_change",
-        current_context=current_context,
-        next_context=next_context
-    )
 
+    _depth = 0
```

Comment:
Does this counter assume that there is only one engine running at one time?
Right now, I don't think there isn't a huge amount of code that assumes this (at least there didn't use to be, it may have changed). I wouldn't rule out having parallel engines running in the future (for plugins for example), so I am trying to understand if this change limits us in that respect.